anvil/sample.py

```diff
@@ -12,18 +12,132 @@
 from reportengine import collect
 
 
-def sample(loaded_model, action, target_length: int, n_large=20000) -> torch.Tensor:
+def chain_autocorrelation(history: torch.Tensor) -> float:
+    r"""
+    Compute the integrated autocorrelation time for the Markov chain.
+
+        \tau_int = 0.5 + sum_{\tau=1}^{\tau_max} \rho(\tau)/\rho(0)
+
+    where \rho(\tau)/\rho(0) is the probability of \tau consecutive rejections,
+    which we estimate by
+
+        \rho(\tau)/\rho(0) = # consecutive runs of \tau rejections / (N - \tau)
+
+    See eqs. (16) and (19) in https://arxiv.org/pdf/1904.12072.pdf
+
+    Parameters
+    ----------
+    history: torch.Tensor
+        A 1xN boolean tensor encoding the acceptance/rejection history of the
+        Markov chain. An accepted (rejected) move is denoted by True (False).
+
+    Returns
+    -------
+    integrated_autocorrelation: float
+        The integrated autocorrelation time for the Markov chain.
+
+    """
+
+    N = len(history)
+    autocorrelations = torch.zeros(N - 1, dtype=torch.float)
+    consecutive_rejections = 0
+
+    for step in history:
+        if step == True:  # move accepted
+            if consecutive_rejections > 0:  # faster than unnecessarily accessing array
+                autocorrelations[1 : consecutive_rejections + 1] += 1
+            consecutive_rejections = 0
+        else:  # move rejected
+            consecutive_rejections += 1
+    if consecutive_rejections > 0:  # pick up last rejection run
+        autocorrelations[1 : consecutive_rejections + 1] += 1
+
+    # Compute integrated autocorrelation
+    integrated_autocorrelation = 0.5 + torch.sum(
+        autocorrelations / (N - torch.arange(N - 1))
+    )
+
+    return integrated_autocorrelation
+
+
+def produce_samples(
+    loaded_model, action, length, state_i=None, log_ratio_i=None, first=False
+):
     r"""
     Sample using Metroplis-Hastings algorithm from a large number of phi
     configurations.
-
+    
```

Comment:
trailing whitespace in general is frowned upon, at the very least it looks sloppy - fine... At the worst I think it can cause genuine problems

e.g https://softwareengineering.stackexchange.com/a/121560

running `pylint` on this file will make it complain, running `black` will just fix it for you (and do other formatting things according to a mixture of PEP and black specific rules

Reply:
ach, I ran black and it's not changing it! Probably need to configure it or something..

Reply:
that's a shame, it's normally pretty good. I'm also running pylint on VS code so this kind of thing gets highlighted to me.

Sorry for being a bit pedantic but a wise person once told me a [story](https://www.insider.com/van-halen-brown-m-ms-contract-2016-9) of why if you see something like whitespace or no final newline in code them you instantly become suspicious that there could be other more serious